### ISSUE TITLE
Feature request: Imgur upload support

### ISSUE THREAD

**Issue by zero-master** [phw/peek#298]
Peek works for me but uploading gif manually is still a pain.

**Comment by luzpaz** [comment]
Code from xfce4-screenshooter to upload to imgur (it's in c) 
https://github.com/xfce-mirror/xfce4-screenshooter/blob/master/lib/screenshooter-imgur.c
https://github.com/xfce-mirror/xfce4-screenshooter/blob/master/lib/screenshooter-imgur.h

**Comment by gort818** [comment]
What would be the best to implement this? Have an check option for imgur upload, when you are done recording open a dialog box asking whether you wish to upload? or just skip the dialog and get straight to the upload?

I might be interested in working on this.. what are your opinions on the best way to go about doing this?

**Comment by luzpaz** [comment]
Both :stuck_out_tongue_closed_eyes:  
dialog box + the option to make it seamless ("don't ask me again" checkbox)
What would be awesome but perhaps overcomplicated:
![proxy duckduckgo com](https://user-images.githubusercontent.com/4140247/37958797-773f25d4-317f-11e8-8fd2-6ae990d3c31d.png)
Just like in xfce4-screenshooter a dialog box that pops up after the screenshot is completed, allows different option, especially the option to open the screenshot in a certain application like gimp, inkscape etc...
In the context of Peek, folks could choose to open the animated gif or mp4 with kdenlive for example...

**Comment by gort818** [comment]
I like it but it does seem to a little too much.. what do you think @phw ?

**Comment by phw** [comment]
I am not sure about this. Maybe do it like this:

1. Have a sharing dialog shown after saving, for now with only the options: 
    - Open in file manager
    - Upload to Imgur
    - Maybe the "open with" as shown above, but we could do this in a separate issue
2. Remove the "Show file manager after saving" option, and replace it with "Show sharing dialog after saving"

Not sure about having a default option. When this bypasses the dialog I don't know how people escape from this again. I would exclude this for now.

For the dialog itself I would keep it very simple and easy to understand. I like the selection boxes Gnome Tweak Tools uses for some choices, they provide some extra information:

![grafik](https://user-images.githubusercontent.com/29852/38074166-22d074a2-332e-11e8-94e0-d77646ec27b9.png)

I am not sure how Imgur works, but I assume you need to authenticate? So we need some functionality for this as well.

**Comment by zero-master** [comment]
These might help you:
https://github.com/3demax/tray-uploader/blob/master/tray.vala
https://github.com/antono/synapse/blob/master/src/plugins/imgur-plugin.vala
https://gist.github.com/twolfson/ad9084fb286e1baee969bbe5eabebe5f

**Comment by gort818** [comment]
@zero-master thanks I will look into this.

**Comment by A1rPun** [comment]
If this will get implemented you only have to wait for the following questions : can it upload to facebook? can it upload to instagram?

The "upload to site X" features rely on a third party site and as a dev I'm not a fan of that. A better way is have another neglectible program that uploads a folder or individual images to any media site.

This is just my 2 cents. Also if I see some program use a third party I don't trust I stop using the program.

**Comment by A1rPun** [comment]
@phw Fair enough, code can be altered and be updated when said services go down (if ever).

I never stated that I don't trust Imgur, I just wanted to protect this great tool for more issue requests for other different services which I don't trust.

**Comment by phw** [comment]
> I never stated that I don't trust Imgur, I just wanted to protect this great tool for more issue requests for other different services which I don't trust.

Definitely. I think you can be sure this won't get too much. But we need to start with some sharing features, I tried ignoring this and similar feature requests long enough :) And Imgur is a) popular enough and b) the API looks simple enough to give this a go.

**Comment by A1rPun** [comment]
Alright good luck with the implementation :)! Imgur is indeed a popular site, I personally use Giphy but can't have it all :stuck_out_tongue:

**Comment by zero-master** [comment]
If you want this tool to be popular, you'll have to support use case which enables easy sharing on at least one of the top website.

What's wrong with imgur?

I always had this tool installed but I never looked for it because I knew, I'll have to manually take the gif and upload it somewhere before I send the URL to my friend. That's too much effort which shouldn't be needed in 2018.

**Comment by gort818** [comment]
I got a working example now we just need to implement it for peek.. it is a start.
```
using Soup;
void main () {
   

   string uri= "https://raw.githubusercontent.com/phw/peek/master/data/screenshots/peek-recording-itself.gif";
   string image = uri;
   
   var mpart = new Multipart(FORM_MIME_TYPE_MULTIPART);
   var session = new Soup.Session ();
   
  
   mpart.append_form_string("image", image);
   var message = Soup.Form.request_new_from_multipart("https://api.imgur.com/3/image",mpart);  
   
   message.request_headers.append("Authorization", "Client-ID bfb3ac58837a0d0");
   session.send_message(message);
   var response = (string) message.response_body.data;
   //print(response);
  
   var parser = new Json.Parser ();
   parser.load_from_data (response, -1);
    
   var root_object = parser.get_root ().get_object ();
        string link = root_object.get_object_member ("data")
                                       .get_string_member ("link");
    print(link);
    
   
}

```

**Comment by gort818** [comment]
@phw I have start the implementation of the imgur upload.. but ran into a problem,

> fatal error: libsoup/soup.h: No such file or directory

I added the package to CMakelists as libsoup-2.4 but no luck.
It is looking for /usr/include/libsoup/soup.h .. but it is installed as /usr/include/libsoup2.4/libsoup/soup.h
So out of curiosity I created a symbolic link and tested it out, this is what I got

> undefined reference to `soup_session_send_message'

Running the above example I created  by itself  compiles with no errors.. I must be missing something in cmake.

Any help would be great !

**Comment by gort818** [comment]
@A1rPun  I see were you are coming from, but I am not using any of their code or libraries it is a simple https POST request.. 

I also checked the Giphy api.. unfortunately there is no upload api available at the moment they said there are working on it https://github.com/Giphy/GiphyAPI/issues/157

**Comment by gort818** [comment]
Good news I have a working example now just to implement it! Soon we shall be able to upload to imgur!

**Comment by gort818** [comment]
@phw something like this?
![sharing](https://user-images.githubusercontent.com/9098121/38786848-63e7df70-40df-11e8-8c30-8362f4df79eb.gif)

**Comment by phw** [comment]
@gort818 Yes, something like this as selection. When putting it into the final dalog I would just put some spacing around. And a checkmark like in the following screen would help:

![grafik](https://user-images.githubusercontent.com/29852/39292895-c583c75c-4937-11e8-8c96-4d4b348181fe.png)

For the buttons I would suggest adding a headerbar with Ok and Cancel button.

**Comment by gort818** [comment]
@phw any idea how to get the check-marks on the right side? I played with it quite a bit couldn't figure it out with adding a button and label for the check-mark. I wonder if the check-mark moves according the theme.

**Comment by phw** [comment]
@gort818 No, not directly. But I would probably look at how Gnome Tweaks does it and copy that. The screenshot above is from there.

**Comment by gort818** [comment]
![imgur](https://user-images.githubusercontent.com/9098121/39400696-7e07cbda-4aea-11e8-8588-7f2b99f82269.gif)

Ok getting somehere

**Comment by gort818** [comment]
Ok I got the basic functionality working if any of you want to give a test drive, atm peek returns the imgur link in the terminal.
https://github.com/gort818/peek/tree/imgur

**Comment by phw** [comment]
Thanks a lot for this! I will definitely give this a test drive in the next days!

**Comment by gort818** [comment]
I have updated my Imgur branch recently if anyone wants to test it out,
https://github.com/gort818/peek/tree/imgur

**Comment by tylerecouture** [comment]
Any update on getting this feature added? =D
